### PR TITLE
Add digital ocean floating ips

### PR DESCRIPTION
**Pull request by gravyboat**
This adds support for floating IPs within Digital Ocean.

### CODE REVIEW

**Comment by gravyboat** [comment]
I've fixed up the whitespace/blank line issues in pylint (will confirm once it finishes), but I'm not really sure how to resolve this correctly so it's compatible with python 2.6: `01:52:37 salt/cloud/clouds/digital_ocean.py:894: [E0598(minimum-python-version), ] Incompatible Python 2.6 code found: set literal` which is located here: https://github.com/saltstack/salt/pull/28194/files#diff-6e82f218b20dac0495868c38492c4c49R894. If anyone has suggestions I'd be happy to hear them.

**Comment by cachedout** [comment]
@gravyboat To me, that Pylint error seems spurious. Here's an example in 2.6 that mimics what you're doing:

```
Python 2.6.9 (unknown, Aug 19 2015, 12:01:23) 
[GCC 5.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> x = {'a': 'b'}
>>> set(x.keys())
set(['a'])
>>> 
```

I suspect we will be OK if you just add a pylint ignore statement. (There are others in the codebase, just poke around for them or let me know if you need help.)

@rallytime Can you please review the cloud-specific code here to make sure everything looks OK to you and then get this merged when it's ready?

File: salt/cloud/clouds/digital_ocean.py [hunk 1]
```diff
@@ -801,6 +801,208 @@ def show_pricing(kwargs=None, call=None):
     return {profile['profile']: ret}
 
 
+def list_floating_ips(call=None):
+    '''
+    Return a list of the floating ips that are on the provider
```
Comment (rallytime) [inline salt/cloud/clouds/digital_ocean.py]: Can you add a `.. versionadded:: Boron` statement here?

File: salt/cloud/clouds/digital_ocean.py [hunk 2]
```diff
@@ -801,6 +801,208 @@ def show_pricing(kwargs=None, call=None):
...
+        except KeyError:
+            fetch = False
+
+    return ret
+
+
+def show_floating_ip(kwargs=None, call=None):
+    '''
+    Show the details of a floating IP
+
```
Comment (rallytime) [inline salt/cloud/clouds/digital_ocean.py]: Same comment about the versionadded directive here as well.

File: salt/cloud/clouds/digital_ocean.py [hunk 3]
```diff
@@ -801,6 +801,208 @@ def show_pricing(kwargs=None, call=None):
...
+    if call != 'function':
+        log.error(
+            'The create_floating_ip function must be called with -f or --function.'
+        )
+        return False
+
+    if not kwargs:
+        kwargs = {}
+
+    if not {'region', 'droplet_id'} & set(kwargs.keys()):
```
Comment (rallytime) [inline salt/cloud/clouds/digital_ocean.py]: Can you change this from `&` to `and`?

**Comment by rallytime** [comment]
Thanks @gravyboat! I added a couple of comments but otherwise this looks good! Once you get those cleaned up we can merge this in. :)

**Comment by gravyboat** [comment]
@rallytime Thanks for the review. I've fixed up those `versionadd` statements so they're present for all the new functions. What do you think about modifying this chunk:
```
    if not {'region', 'droplet_id'} and set(kwargs.keys()):
        log.error('A droplet_id or region is required.')
        return False

    if 'droplet_id' in kwargs:
        result = query(method='floating_ips',
                           args={'droplet_id': kwargs['droplet_id']},
                           http_method='post')

    elif 'region' in kwargs:
        result = query(method='floating_ips',
                           args={'region': kwargs['region']},
                           http_method='post')

    return result
```
To this for 2.6 support:
```
    if 'droplet_id' in kwargs:
        result = query(method='floating_ips',
                           args={'droplet_id': kwargs['droplet_id']},
                           http_method='post')

        return result

    elif 'region' in kwargs:
        result = query(method='floating_ips',
                           args={'region': kwargs['region']},
                           http_method='post')

        return result

    else:
        log.error('A droplet_id or region is required.')
        return False
```

Does that go against convention?

**Comment by rallytime** [comment]
@gravyboat Nope, not at all. That looks great!

**Comment by gravyboat** [comment]
@rallytime I think this is all set to go. None of the failures I see are related to this PR.

**Comment by rallytime** [comment]
Awesome! Thanks @gravyboat!

**Pull request merged**